assembly/src/parser/grammar.lalrpop

```diff
@@ -0,0 +1,1378 @@
+    "u32cto" => Instruction::U32Cto,
+    "u32split" => Instruction::U32Split,
+    "u32test" => Instruction::U32Test,
+    "u32testw" => Instruction::U32TestW,
+    "u32wrapping_add3" => Instruction::U32WrappingAdd3,
+    "u32wrapping_madd" => Instruction::U32WrappingMadd,
+    "u32xor" => Instruction::U32Xor,
+    "xor" => Instruction::Xor,
+}
+
+#[inline]
+AdviceInjector: Instruction = {
+    "adv.insert_hdword" <domain:MaybeImm<U8>> => {
+        domain.map(|domain| Instruction::AdvInject(AdviceInjectorNode::InsertHdwordImm { domain }))
+              .unwrap_or(Instruction::AdvInject(AdviceInjectorNode::InsertHdword))
+    },
+    "adv.insert_hperm" => Instruction::AdvInject(AdviceInjectorNode::InsertHperm),
+    "adv.insert_mem" => Instruction::AdvInject(AdviceInjectorNode::InsertMem),
+    "adv.push_ext2intt" => Instruction::AdvInject(AdviceInjectorNode::PushExt2intt),
+    "adv.push_mapval" <i:MaybeImm<StackWordOffset>> => {
+        i.map(|offset| Instruction::AdvInject(AdviceInjectorNode::PushMapValImm { offset }))
+         .unwrap_or(Instruction::AdvInject(AdviceInjectorNode::PushMapVal))
+    },
+    "adv.push_mapvaln" <i:MaybeImm<StackWordOffset>> => {
+        i.map(|offset| Instruction::AdvInject(AdviceInjectorNode::PushMapValNImm { offset }))
+         .unwrap_or(Instruction::AdvInject(AdviceInjectorNode::PushMapValN))
+    },
+    "adv.push_mtnode" => Instruction::AdvInject(AdviceInjectorNode::PushMtNode),
+    "adv.push_sig" "." <kind:SignatureKind> => Instruction::AdvInject(AdviceInjectorNode::PushSignature { kind }),
+    "adv.push_smtpeek" => Instruction::AdvInject(AdviceInjectorNode::PushSmtPeek),
+    "adv.push_smtget" => Instruction::AdvInject(AdviceInjectorNode::PushSmtGet),
+    "adv.push_smtset" => Instruction::AdvInject(AdviceInjectorNode::PushSmtSet),
+    "adv.push_u64div" => Instruction::AdvInject(AdviceInjectorNode::PushU64Div),
+}
+
+#[inline]
+Assert: Instruction = {
+    "assert" <error_code:MaybeAssertCode> => error_code.map(Instruction::AssertWithError).unwrap_or(Instruction::Assert),
+    "assertz" <error_code:MaybeAssertCode> => error_code.map(Instruction::AssertzWithError).unwrap_or(Instruction::Assertz),
+    "assert_eq" <error_code:MaybeAssertCode> => error_code.map(Instruction::AssertEqWithError).unwrap_or(Instruction::AssertEq),
+    "assert_eqw" <error_code:MaybeAssertCode> => error_code.map(Instruction::AssertEqwWithError).unwrap_or(Instruction::AssertEqw),
+    "u32assert" <error_code:MaybeAssertCode> => error_code.map(Instruction::U32AssertWithError).unwrap_or(Instruction::U32Assert),
+    "u32assert2" <error_code:MaybeAssertCode> => error_code.map(Instruction::U32Assert2WithError).unwrap_or(Instruction::U32Assert2),
+    "u32assertw" <error_code:MaybeAssertCode> => error_code.map(Instruction::U32AssertWWithError).unwrap_or(Instruction::U32AssertW),
+}
+
+MaybeAssertCode: Option<Immediate<u32>> = {
+    "." "err" "=" <code:ImmValue<U32>> => Some(code),
+    => None,
+}
```

Comment:
I think we should forbid `assert.err=0`, because `0` is used as the default for a plain `assert`. An alternative would be to have the error in the VM be a `Option<u32>` instead. So the default is set to `None` instead.

My rationale is that these two error spaces should not collide.

Reply:
Yeah, totally agree. Ideally we would have some stack tracing too. I think the issue with that is that MAST erases the procedure names? But yeah, I'm 100% with you, would love to have additional information for errors.

Reply:
One of the main motivations for limiting error codes to 32 bits is to keep the size of programs relatively small which is important for putting programs on chain. Otherwise, with this new structure, nothing really prevents us from using strings or more elaborate identifiers. A possible strategy to adopt is to have more elaborate identifiers but then strip them out for on-chain programs during serialization. But that's more related to MAST serialization than this PR.

Reply:
The `SourceSpan` type is 32 bits, but in order to be useful, we also need to have the source code (path or actual content), so that we can render an error with a source snippet (see the assembler tests for examples of what these look like if you haven't seen them yet). We don't actually need identifiers (file/line/col) at each site. The source code path (or content) could be provided with the MAST by mapping each MAST root corresponding to a procedure with the source code for that procedure when built with debug info, allowing us to render source code snippets by looking up the source for the current procedure (if available) and then rendering the span using the source span encoded in the instruction.

Of course, the problem occurs when the source code is _not_ available, still, the source span value itself forms a useful and semi-unique error code (i.e. identical spans with different errors are possible, since spans are unique only within a file, but they probably aren't super common).

If we used 64 bits for errors, then 32 bits could be the user-provided code, and 32 bits could be the span. We could even omit the 32 bits used for custom codes when no code is used, and so save space in the MAST. The main thing we'd need to do though is ensure the span is not part of the hash, so as to avoid making two otherwise identical functions have unique hashes. It's not a huge issue if the span ends up in the hash, but it would preclude two functions defined separately from hashing to the same value (unless they just happened to occur in the same place in both files, but that's super unlikely).